[4.0] RavenDB-10008 Go to real leader for cluster topology

karmeli87 opened the pull request.

karmeli87 committed:
RavenDB-10008 Go to real leader for cluster topology

ayende commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -170,6 +171,55 @@ public long LastRaftCommitIndex
+                }
+                catch (Exception)
+                {
+                    //
```
We should at least have logs here. And why is this an ongoing operation?

ayende commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -170,6 +171,55 @@ public long LastRaftCommitIndex
+            {
+                try
+                {
+                    if (await _engine.WaitForState(RachisState.Follower).WaitWithTimeout(TimeSpan.FromSeconds(10)) == false)
```
Why every ten seconds?

ayende commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -170,6 +171,55 @@ public long LastRaftCommitIndex
+                        using (ContextPool.AllocateOperationContext(out JsonOperationContext context))
+                        {
+                            var leaderWsUrl = new Uri($"{leaderUrl.Replace("http", "ws", StringComparison.OrdinalIgnoreCase)}/server/notification-center/watch");
+                            await ws.ConnectAsync(leaderWsUrl, cts.Token);
```
Missing certificate usage?

ayende commented on src/Raven.Server/ServerWide/ServerStore.cs:
```diff
@@ -509,6 +559,7 @@ public void TriggerDatabases()
             }
 
             Task.Run(ClusterMaintenanceSetupTask, ServerShutdown);
+            Task.Run(UpdateTopologyChangeNotification, ServerShutdown);
```
There is no point in running this unless we have the studio open

karmeli87 committed:
RavenDB-10008 Go to real leader for cluster topology

PR fixes

karmeli87 committed:
RavenDB-10008 Go to real leader for cluster topology

PR fix, if the websocket was closed don't wait on the cancelTask, but check if it completed.
